### PR TITLE
prefer IR over ApplyIR for simple functions

### PR DESCRIPTION
**Pull request by cseed**
improves optimization and opportunities for deforestation

### CODE REVIEW

File: hail/python/hail/expr/functions.py [hunk 1]
```diff
@@ -2959,8 +2959,10 @@ def len(x) -> Int32Expression:
     """
     if isinstance(x.dtype, ttuple) or isinstance(x.dtype, tstruct):
         return hl.int32(builtins.len(x))
+    elif x.dtype == tstr:
+        return apply_expr(lambda x: StringLength(x), tint32, x)
     else:
-        return x._method("size", tint32)
+        return apply_expr(lambda x: ArrayLen(x), tint32, x)
```
Comment (tpoterba) [inline hail/python/hail/expr/functions.py]: does this work with sets? Don't they need to be `ToArray`ed first?

Comment (catoverdrive) [reply hail/python/hail/expr/functions.py]: I think they do. (Same thing applies for `CollectionExpression.size()` and `.length()` above)

**Comment by cseed** [comment]
Yep, I'll get to this soon.

File: hail/python/hail/expr/functions.py (hunk 1 above)
Comment (cseed) [reply hail/python/hail/expr/functions.py]: Should it?

**Comment by cseed** [comment]
@catoverdrive I see now there is an unwrap that should have had this same effect, so I don't quite understand why this was necessary (although it seemed to be).  Anyway, this still seems simpler/more correct.

**Pull request merged**